### PR TITLE
Parsing performance improvements (not for direct merge)

### PR DESCRIPTION
**Pull request by sixlettervariables**
This PR is for review/comments only and should probably be brought in in parts, depending on the goals of the project (readability vs performance).

General strategy:
1. Reduce allocations.
2. Pool allocations.
3. Reduce LINQ usage.

BenchmarkDotNet results for the PR compared to 9.5.X and Newtonsoft.Json 10.X (read the commit messages for their individual improvements):
```
BenchmarkDotNet=v0.10.11, OS=Windows 7 SP1 (6.1.7601.0)
Processor=Intel Xeon CPU E5-2620 v3 2.40GHzIntel Xeon CPU E5-2620 v3 2.40GHz, ProcessorCount=24
Frequency=2338369 Hz, Resolution=427.6485 ns, Timer=TSC
  [Host]     : .NET Framework 4.7 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2114.0
  DefaultJob : .NET Framework 4.7 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2114.0

                            Method |     Mean |     Error |    StdDev |     Gen 0 |    Gen 1 |    Gen 2 | Allocated |
---------------------------------- |---------:|----------:|----------:|----------:|---------:|---------:|----------:|
 JsonSerializer_Deserialize_Stream | 38.16 ms | 0.9832 ms | 2.8990 ms | 1062.5000 | 437.5000 | 125.0000 |   5.76 MB |
                     JObject_Parse | 28.57 ms | 0.8978 ms | 2.6470 ms | 1180.0000 | 495.9375 | 159.6875 |   6.26 MB |
                     
            JsonValue_Parse_Stream | 54.46 ms | 1.7220 ms | 5.0770 ms | 4062.5000 | 625.0000 | 187.5000 |  23.54 MB |
            JsonValue_Parse_String | 42.52 ms | 1.0770 ms | 3.1740 ms | 4062.5000 | 625.0000 | 187.5000 |  23.54 MB |
              JsonValue_ParseAsync | 66.57 ms | 1.3300 ms | 3.2630 ms | 4062.5000 | 625.0000 | 187.5000 |  23.55 MB |
              
        New_JsonValue_Parse_Stream | 27.34 ms | 0.9102 ms | 2.6836 ms |  625.0000 | 312.5000 |        - |   3.96 MB |
        New_JsonValue_Parse_String | 12.49 ms | 0.2479 ms | 0.2198 ms |  687.5000 | 312.5000 |  62.5000 |   3.95 MB |
          New_JsonValue_ParseAsync | 28.94 ms | 0.7739 ms | 2.2697 ms |  656.2500 | 312.5000 |  62.5000 |   3.96 MB |
```

I also need to fix-up my whitespace to match yours.

### CODE REVIEW

File: Manatee.Json/Parsing/BoolParser.cs [hunk 1]
```diff
@@ -8,102 +8,167 @@ namespace Manatee.Json.Parsing
...
-			var result = new string(buffer).ToLower();
-			if (result == "true")
+			value = null;
+
+			if (source[index] == 't' || source[index] == 'T')
 			{
+				if (index + 4 >= source.Length)
+					return UnexpectedEndOfInput;
+
+				if (source.IndexOf("true", index, 4, StringComparison.OrdinalIgnoreCase) != index)
```
Comment (gregsdennis) [inline Manatee.Json/Parsing/BoolParser.cs]: JSON specifies that true/false should be lower-case.  I don't mind having an option to ignore case, but I'd like the default to be conformant.

File: Manatee.Json/Parsing/NullParser.cs [hunk 2]
```diff
@@ -7,58 +7,79 @@ namespace Manatee.Json.Parsing
...
-			{
-				value = null;
-				return $"Value not recognized: '{result}'.";
-			}
+			value = null;
+
+			if (index + 4 >= source.Length)
+				return UnexpectedEndOfInput;
+
+			if (source.IndexOf("null", index, 4, System.StringComparison.OrdinalIgnoreCase) != index)
```
Comment (gregsdennis) [inline Manatee.Json/Parsing/NullParser.cs]: Again, `null` is defined as being strictly lower-case.

**Comment by gregsdennis** [comment]
Overall, I like what you've done.  I'm impressed with the improvements around reorganizing the individual parser logic, and I like the StringBuilder and buffer caches.

I understand the need for some of the changes, like removing `.In()` and using switches instead.  It's not as pretty, but I think not having to convert parameters into an array then iterating that array is probably more important.

I'd like to pull this into a branch and make some stylistic changes, but I think the functional part is good.

File: Manatee.Json/Parsing/BoolParser.cs (hunk 1 above)
Comment (sixlettervariables) [reply Manatee.Json/Parsing/BoolParser.cs]: I was following the original logic which did a `ToLower()` before comparing to `"true"` or `"false"`.

Comment (gregsdennis) [reply Manatee.Json/Parsing/BoolParser.cs]: fair enough.  I think I'll change that logic and move it to an option.

**Comment by sixlettervariables** [comment]
Sounds good. The caching comes from [Roslyn](http://mattwarren.org/2014/06/10/roslyn-code-base-performance-lessons-part-2/), and should be cleaned up for inclusion. I'm not sure all of the defaults are appropriate.

I'm now taking a look at the serialization/deserialization to object side.

**Comment by gregsdennis** [comment]
@sixlettervariables I appreciate the effort.  Serialization has become something of a rat's nest of attempts to cache reflection and architecture out the different serialization methods.  That's why I haven't gone back in to tackle it yet.

**Comment by gregsdennis** [comment]
@sixlettervariables I see 13 parsing errors in the tests.  Most of them have to do with unicode escape sequences, but a few deal with parsing lone explicit values (`null`, `true`, & `false`).

**Comment by gregsdennis** [comment]
@sixlettervariables please pull my changes from https://github.com/gregsdennis/Manatee.Json/tree/task/issue-9-parsing-perf-improvements.  I have all of the tests passing.

**Comment by sixlettervariables** [comment]
Awesome! I'll update now.

**Comment by gregsdennis** [comment]
Closing this as we have a local branch managing these changes.

**Pull request closed without merging**